### DISCUSSION TITLE
Different cluster node names

### DISCUSSION THREAD

**Discussion by mxmcarl** [rabbitmq/rabbitmq-server#3114] [Other] [answered]
Hi, 
    I built a cluster of three nodes in Erlang/OTP 24 and Rabbitmq V3.8.17.But they have different cluster names.The ways to build clusters are peer discovery(rabbitmq_peer_discovery_common Plugin)

[root@host1 ~]# rabbitmqctl eval "rabbit_mnesia:cluster_nodes(all)."
['rabbit@host3.localdomain','rabbit@host2.localdomain',
 'rabbit@host1.localdomain']
[root@host1 ~]# rabbitmqctl eval "rabbit_nodes:cluster_name()."
<<"rabbit@host1.localdomain">>
[root@host1 ~]# 

[root@host2 ~]# rabbitmqctl eval "rabbit_mnesia:cluster_nodes(all)."
['rabbit@host3.localdomain','rabbit@host2.localdomain',
 'rabbit@host1.localdomain']
[root@host2 ~]# rabbitmqctl eval "rabbit_nodes:cluster_name()."
<<"rabbit@host2.localdomain">>
[root@host2 ~]# 

[root@host3 ~]# rabbitmqctl eval "rabbit_mnesia:cluster_nodes(all)."
['rabbit@host3.localdomain','rabbit@host2.localdomain',
 'rabbit@host1.localdomain']
[root@host3 ~]# rabbitmqctl eval "rabbit_nodes:cluster_name()."
<<"rabbit@host3.localdomain">>
[root@host3 ~]# 

 [root@host3 ~]# rabbitmqctl cluster_status
Cluster status of node rab...@host3.localdomain ...
Basics

Cluster name: rab...@host3.localdomain

Disk Nodes

rab...@host1.localdomain
rab...@host2.localdomain
rab...@host3.localdomain

Running Nodes

rab...@host1.localdomain
rab...@host2.localdomain
rab...@host3.localdomain

Versions

rab...@host1.localdomain: RabbitMQ 3.8.17 on Erlang 24.0.2
rab...@host2.localdomain: RabbitMQ 3.8.17 on Erlang 24.0.2
rab...@host3.localdomain: RabbitMQ 3.8.17 on Erlang 24.0.2

Maintenance status

Node: rab...@host1.localdomain, status: not under maintenance
Node: rab...@host2.localdomain, status: not under maintenance
Node: rab...@host3.localdomain, status: not under maintenance

Alarms

(none)

Network Partitions

(none)

Listeners

Node: rab...@host1.localdomain, interface: [::], port: 15672, protocol: http, purpose: HTTP API
Node: rab...@host1.localdomain, interface: [::], port: 61613, protocol: stomp, purpose: STOMP
Node: rab...@host1.localdomain, interface: [::], port: 15674, protocol: http/web-stomp, purpose: STOMP over WebSockets
Node: rab...@host1.localdomain, interface: [::], port: 1883, protocol: mqtt, purpose: MQTT
Node: rab...@host1.localdomain, interface: [::], port: 25672, protocol: clustering, purpose: inter-node and CLI tool communication
Node: rab...@host1.localdomain, interface: [::], port: 5672, protocol: amqp, purpose: AMQP 0-9-1 and AMQP 1.0
Node: rab...@host2.localdomain, interface: [::], port: 15674, protocol: http/web-stomp, purpose: STOMP over WebSockets
Node: rab...@host2.localdomain, interface: [::], port: 25672, protocol: clustering, purpose: inter-node and CLI tool communication
Node: rab...@host2.localdomain, interface: [::], port: 15672, protocol: http, purpose: HTTP API
Node: rab...@host2.localdomain, interface: [::], port: 61613, protocol: stomp, purpose: STOMP
Node: rab...@host2.localdomain, interface: [::], port: 1883, protocol: mqtt, purpose: MQTT
Node: rab...@host2.localdomain, interface: [::], port: 5672, protocol: amqp, purpose: AMQP 0-9-1 and AMQP 1.0
Node: rab...@host3.localdomain, interface: [::], port: 15674, protocol: http/web-stomp, purpose: STOMP over WebSockets
Node: rab...@host3.localdomain, interface: [::], port: 25672, protocol: clustering, purpose: inter-node and CLI tool communication
Node: rab...@host3.localdomain, interface: [::], port: 15672, protocol: http, purpose: HTTP API
Node: rab...@host3.localdomain, interface: [::], port: 61613, protocol: stomp, purpose: STOMP
Node: rab...@host3.localdomain, interface: [::], port: 1883, protocol: mqtt, purpose: MQTT
Node: rab...@host3.localdomain, interface: [::], port: 5672, protocol: amqp, purpose: AMQP 0-9-1 and AMQP 1.0

Feature flags

Flag: drop_unroutable_metric, state: disabled
Flag: empty_basic_get_metric, state: disabled
Flag: implicit_default_bindings, state: enabled
Flag: maintenance_mode_status, state: enabled
Flag: quorum_queue, state: enabled
Flag: user_limits, state: enabled
Flag: virtual_host_metadata, state: enabled

**Comment by michaelklishin** [comment]
You can explicitly set cluster name to any value you like. By default it is "computed" from the name of the node itself:

``` erl
-spec cluster_name() -> binary().

cluster_name() ->
    case rabbit_runtime_parameters:value_global(cluster_name) of 
        not_found -> cluster_name_default();
        Name -> Name
    end.

cluster_name_default() ->
    {ID, _} = parts(node()),
    FQDN = rabbit_net:hostname(),
    list_to_binary(atom_to_list(make({ID, FQDN}))).
```

so when a cluster is formed in parallel every node will have its own value. The name is not used by RabbitMQ except for some entities declared by Federation links, so I don't see any issue with it. There is no way to solve this without requiring a sequential node
startup (one by one).

**Comment by michaelklishin** [comment, marked as answer]
A couple more things are worth mentioning. You can explicitly set cluster ID [via `rabbitmq.conf`](https://github.com/rabbitmq/rabbitmq-server/blob/v3.8.x/deps/rabbit/docs/rabbitmq.conf.example#L253-L255) which would make it available to every cluster member even very early on boot. Perhaps that's what you want here, instead of using CLI tools to set cluster name.

Also, nodes have an internal cluster ID that's not exposed to the user. That is used by Tanzu RabbitMQ (the commercial edition) and are not exposed to the user in any way. So when a node needs a unique cluster ID that's consistent, there is such mechanism and it cannot be changed by the users.

**Reply by mxmcarl** [reply]
OK， I‘ll set cluster ID via rabbitmq.conf.
When cluster ID is needed, internal ID is used instead of user set ID, right？

**Reply by michaelklishin** [reply]
The cluster ID you control is for your own convenience and as an extra identifier in e.g. federation link consumer tags.
AMQP 1.0 plugin uses it as a container ID value, too.

Unless your deployment tools depend on it one way or another, it's very unlikely that changing it will have any visible effects.
